src/heartbeats.cpp

```diff
@@ -0,0 +1,611 @@
+/**
+ * @file heartbeats.cpp
+ * @brief Classes for heartbeating Sync configuration and status
+ *
+ * (c) 2013 by Mega Limited, Auckland, New Zealand
+ *
+ * This file is part of the MEGA SDK - Client Access Engine.
+ *
+ * Applications using the MEGA API must present a valid application key
+ * and comply with the the rules set forth in the Terms of Service.
+ *
+ * The MEGA SDK is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * @copyright Simplified (2-clause) BSD License.
+ *
+ * You should have received a copy of the license along with this
+ * program.
+ */
+
+#include "mega/heartbeats.h"
+#include "mega/command.h"
+#include "assert.h"
+
+namespace mega {
+
+
+HeartBeatSyncInfo::HeartBeatSyncInfo(int tag, handle id)
+    : mHeartBeatId(id), mSyncTag(tag)
+{
+}
+
+handle HeartBeatSyncInfo::heartBeatId() const
+{
+    return mHeartBeatId;
+}
+
+Command *HeartBeatSyncInfo::runningCommand() const
+{
+    return mRunningCommand;
+}
+
+void HeartBeatSyncInfo::setRunningCommand(Command *runningCommand)
+{
+    mRunningCommand = runningCommand;
+}
+
+HeartBeatSyncInfo::Status HeartBeatSyncInfo::status() const
+{
+    return mStatus;
+}
+
+double HeartBeatSyncInfo::progress() const
+{
+    return std::max(0., std::min(1., static_cast<double>(mTransferredBytes) / static_cast<double>(mTotalBytes)));
+}
+
+uint32_t HeartBeatSyncInfo::pendingUps() const
+{
+    return mPendingUps;
+}
+
+uint32_t HeartBeatSyncInfo::pendingDowns() const
+{
+    return mPendingDowns;
+}
+
+m_time_t HeartBeatSyncInfo::lastAction() const
+{
+    return mLastAction;
+}
+
+mega::MegaHandle HeartBeatSyncInfo::lastSyncedItem() const
+{
+    return mLastSyncedItem;
+}
+
+void HeartBeatSyncInfo::updateTransferInfo(MegaTransfer *transfer)
+{
+    if (mPendingTransfers.find(transfer->getTag()) == mPendingTransfers.end())
+    {
+        mPendingTransfers.insert(std::make_pair(transfer->getTag(), ::mega::make_unique<PendingTransferInfo>()));
+    }
+
+    const auto &pending = mPendingTransfers[transfer->getTag()];
+
+    auto total = mTotalBytes;
+    auto transferred = mTransferredBytes;
+
+    // reduce globals by the last known data
+    total -= pending->totalBytes;
+    transferred -= pending->transferredBytes;
+
+    // update values with those of the transfer
+    pending->totalBytes = transfer->getTotalBytes();
+    pending->transferredBytes = transfer->getTransferredBytes();
+
+    // reflect those in the globals
+    total += pending->totalBytes;
+    transferred += pending->transferredBytes;
+
+    setTotalBytes(total);
```

Comment:
Once the transfer has started (which will represent the first call to this method), does the `totalBytes` of the `transfer` changes? I would not expect so. Maybe for folder-transfers (which we probably want to discard in the heartbeats, since we have subtransfers for each.

Reply:
Neither do I, but I'd rather not count on that: in case we ever use some fancy transfer whose size is not fixed (something streamed), I don't expect high consumption here. Your point is very valid, but I'd rather be conservative here ;-)